tracer/src/Datadog.Trace/Debugger/Instrumentation/AsyncMethodDebuggerInvoker.cs

```diff
@@ -0,0 +1,367 @@
+        /// </summary>
+        private static readonly AsyncLocal<AsyncMethodDebuggerState> AsyncContext = new();
+
+        /// <summary>
+        /// Create a new context and save the parent.
+        /// We call this method from the async kick off method before the state machine is created,
+        /// so for each async state machine, this method will be called exactly once.
+        /// The IsFirstEntry property of the AsyncMethodDebuggerState will be "true"
+        /// then, in the first state machine step, we set this property to "false", thus
+        /// we can deduce that any further invocation is not a new logical call but rather a
+        /// continuation from an "await" expression.
+        /// </summary>
+        /// <param name="kickoffInfo">The async kickoff method info</param>
+        /// <returns>Live debugger state</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static AsyncMethodDebuggerState SetContext(AsyncHelper.AsyncKickoffMethodInfo kickoffInfo)
+        {
+            var currentState = AsyncContext.Value;
+            var newState = new AsyncMethodDebuggerState
+            {
+                Parent = currentState,
+                KickoffInvocationTarget = kickoffInfo.KickoffParentObject,
+                KickoffInvocationTargetType = kickoffInfo.KickoffParentType,
+                KickoffMethod = kickoffInfo.KickoffMethod
+            };
+            AsyncContext.Value = newState;
+            return newState;
+        }
+
+        /// <summary>
+        /// Restore the async context to the parent
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void RestoreContext()
+        {
+            AsyncContext.Value = AsyncContext.Value.Parent;
+        }
+
+        /// <summary>
+        /// Begin Method Invoker
+        /// </summary>
+        /// <typeparam name="TTarget">Target object of the method. Note that it could be typeof(object) and not a concrete type</typeparam>
+        /// <param name="probeId">The id of the probe</param>
+        /// <param name="instance">Instance value</param>
+        /// <param name="methodHandle">The handle of the executing method</param>
+        /// <param name="typeHandle">The handle of the type</param>
+        /// <param name="methodMetadataIndex">The index used to lookup for the <see cref="MethodMetadataInfo"/> associated with the executing method</param>
+        /// <param name="isReEntryToMoveNext">If it the first entry to the state machine MoveNext method</param>
+        /// <returns>Live debugger state</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static AsyncMethodDebuggerState BeginMethod<TTarget>(string probeId, TTarget instance, RuntimeMethodHandle methodHandle, RuntimeTypeHandle typeHandle, int methodMetadataIndex, ref bool isReEntryToMoveNext)
+        {
+            if (ProbeRateLimiter.Instance.IsLimitReached)
```

Comment:
Omer's Probe Rate Limiter PR entered into master. You should rebase and use the new rate limiter like here:
https://github.com/DataDog/dd-trace-dotnet/blob/0290489819371d8b3808cf532ab3e7c084a5e469/tracer/src/Datadog.Trace/Debugger/Instrumentation/MethodDebuggerInvoker.cs#L40